core/src/socket.rs

```diff
@@ -111,100 +160,218 @@ impl<'gc> Sockets<'gc> {
+                    match target {
+                        SocketKind::Avm2(target) => {
+                            let mut activation = Avm2Activation::from_nothing(context.reborrow());
+
+                            let bytes_loaded = data.len();
+                            target.read_buffer().extend(data);
+
+                            let progress_evt = activation
+                                .avm2()
+                                .classes()
+                                .progressevent
+                                .construct(
+                                    &mut activation,
+                                    &[
+                                        "socketData".into(),
+                                        false.into(),
+                                        false.into(),
+                                        bytes_loaded.into(),
+                                        //NOTE: bytesTotal is not used by socketData event.
+                                        0.into(),
+                                    ],
+                                )
+                                .expect("ProgressEvent should be constructed");
+
+                            Avm2::dispatch_event(
+                                &mut activation.context,
+                                progress_evt,
+                                target.into(),
+                            );
+                        }
+                        SocketKind::Avm1(target) => {
+                            let mut activation = Avm1Activation::from_stub(
+                                context.reborrow(),
+                                ActivationIdentifier::root("[XMLSocket]"),
+                            );
+
+                            // NOTE: This is enforced in connect_avm1() function.
+                            let xml_socket =
+                                XmlSocket::cast(target.into()).expect("target should be XmlSocket");
+
+                            let mut buffer = xml_socket.read_buffer();
+                            buffer.extend(data);
+
+                            // Check for a message.
+                            if let Some((index, _)) =
+                                buffer.iter().enumerate().find(|(_, &b)| b == 0)
+                            {
+                                let message = buffer.drain(..index).collect::<Vec<_>>();
+                                // Remove null byte.
+                                let _ = buffer.drain(..1);
+
+                                let message = AvmString::new_utf8_bytes(activation.gc(), &message);
+
+                                let _ = target.call_method(
+                                    "onData".into(),
+                                    &[message.into()],
+                                    &mut activation,
+                                    ExecutionReason::Special,
+                                );
+                            }
```

Comment:
If more than 1 null byte arrives in the same message, you'll miss it until the next message from the server. In my change request I didn't had this issue because I was polling the socket back-end, I did it that way to have at most 1 message fired per update, but in your case this code runs only once per chunk of data received so you risk to miss some messages in some racy conditions.

```suggestion
                            let mut buffer = xml_socket.read_buffer();
                            buffer.extend(data);

                            // Check for a message.
                            while let Some((index, _)) =
                                buffer.iter().enumerate().find(|(_, &b)| b == 0)
                            {
                                let message = buffer.drain(..index).collect::<Vec<_>>();
                                // Remove null byte.
                                let _ = buffer.drain(..1);

                                let message = AvmString::new_utf8_bytes(activation.gc(), &message);

                                let _ = target.call_method(
                                    "onData".into(),
                                    &[message.into()],
                                    &mut activation,
                                    ExecutionReason::Special,
                                );
                            }
```

Reply:
Good catch!

Reply:
Fixed in the latest commit.